### PR TITLE
[Merged by Bors] - Register validator api

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

Lays the groundwork for builder API changes by implementing the beacon-API's new `register_validator` endpoint

## Proposed Changes

- Add a routine in the VC that runs on startup (re-try until success), once per epoch or whenever `suggested_fee_recipient` is updated, signing `ValidatorRegistrationData` and sending it to the BN.
  -  TODO: `gas_limit` config options https://github.com/ethereum/builder-specs/issues/17
-  BN only sends VC registration data to builders on demand, but VC registration data *does update* the BN's prepare proposer cache and send an updated fcU to  a local EE. This is necessary for fee recipient consistency between the blinded and full block flow in the event of fallback.  Having the BN only send registration data to builders on demand gives feedback directly to the VC about relay status. Also, since the BN has no ability to sign these messages anyways (so couldn't refresh them if it wanted), and validator registration is independent of the BN head, I think this approach makes sense. 
- Adds upcoming consensus spec changes for this PR https://github.com/ethereum/consensus-specs/pull/2884
  -  I initially applied the bit mask based on a configured application domain.. but I ended up just hard coding it here instead because that's how it's spec'd in the builder repo. 
  -  Should application mask appear in the api?

### CODE REVIEW

File: beacon_node/http_api/tests/tests.rs
Comment (realbigsean) [inline beacon_node/http_api/tests/tests.rs]:
```suggestion
async fn post_validator_register_validator() {
```

File: consensus/types/src/chain_spec.rs
Comment (realbigsean) [inline consensus/types/src/chain_spec.rs]:
```suggestion
    // This should be updated to include the current fork and the genesis validators root, but discussion is ongoing:
    //
    // https://github.com/ethereum/builder-specs/issues/14
```

File: consensus/types/src/chain_spec.rs
Comment (realbigsean) [inline consensus/types/src/chain_spec.rs]: Update this to actually do the bit mask operation

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs]:
```suggestion
    // Used to track unpublished validator registration changes.
```

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs:137]:
```suggestion
}

```

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs]: Should probably sign on startup, but an epoch before merge seems reasonable to start submitting.. although I guess we don't want to flood the relays at once..

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs]:
```suggestion
                if !guard.contains_key(key) {
```

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs:417]: Should probably only publish unchanged keys so we don't publish twice

File: validator_client/src/signing_method/web3signer.rs
Comment (realbigsean) [inline validator_client/src/signing_method/web3signer.rs:20]: Can I add a web3signer test for this?

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs:216]: Only start this if we have the `private-tx-proposal` flag

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [inline validator_client/src/preparation_service.rs:292]: I think making this configurable should be done in a follow up PR in order to unblock the larger mev-boost compatibility PR

File: validator_client/src/preparation_service.rs
Comment (realbigsean) [reply validator_client/src/preparation_service.rs]: Since we should only publish unchanged signatures once per epoch, I think it's ok to remove this check. And I don't think people will use the builder flow until closer to the merge anyways. I think this is better than potentially flooding the relay an epoch before the merge

File: validator_client/src/signing_method/web3signer.rs
Comment (realbigsean) [reply validator_client/src/signing_method/web3signer.rs:20]: I added tests, but have them commented now until the consensys web3signer adds the `register_validator` method

**Comment by realbigsean** [comment]
Ok this one's ready for review, it will conflict with https://github.com/sigp/lighthouse/pull/3213 but the conflicts should be pretty easy to resolve, happy to merge that one first and fix the conflicts here

File: beacon_node/http_api/src/lib.rs
Comment (paulhauner) [inline beacon_node/http_api/src/lib.rs]:
```suggestion
                    let current_epoch = chain
                        .slot_clock
                        .now_or_genesis()
                        .ok_or(BeaconChainError::UnableToReadSlot)
                        .map_err(warp_utils::reject::beacon_chain_error)?
                        .epoch(T::EthSpec::slots_per_epoch());
```

Otherwise will return an error when run before genesis. We can use `now_or_genesis` to assume the genesis slot before the genesis time.

I doubt it'll ever present as an actual problem, but sometimes we find these issues when waiting for testnets to start and get Discord questions about it.

File: beacon_node/http_api/src/lib.rs
Comment (paulhauner) [inline beacon_node/http_api/src/lib.rs]: I'm interested in the motivation behind this. This would put validator IP info in our `DEBG` logs, which I don't think we usually do. I'm not against it if it's going to be useful.

File: beacon_node/http_api/src/lib.rs
Comment (paulhauner) [inline beacon_node/http_api/src/lib.rs:2486]:
```suggestion

debug!(
    log,
    "Resolved validator request pubkeys";
    "count" => preparation_data.len()
);

```

This might help with debugging so we can detect when people are giving us pubkeys that don't map to a validator index. Just a suggestion.

File: validator_client/src/signing_method.rs
Comment (paulhauner) [inline validator_client/src/signing_method.rs:191]: I'm wondering if we've lost this logic on purpose? I don't think we ever actually use `Deposit`, but it might be nice to have these here for future protections.

File: validator_client/src/preparation_service.rs
Comment (paulhauner) [inline validator_client/src/preparation_service.rs]: Since `RequireSynced::No` we might get an edge-case where if the first BN isn't synced we'll publish these registrations to that one, rather that higher-indexed ones which we're actually using.

Perhaps `RequiredSynced::Yes` is safer?

File: consensus/types/src/application_domain.rs
Comment (paulhauner) [inline consensus/types/src/application_domain.rs]: It might be nice to have `16777216` as a constant so we don't repeat it in other places of the code. Might even be able to make this function a `const fn`?

File: consensus/types/src/chain_spec.rs
Comment (paulhauner) [inline consensus/types/src/chain_spec.rs]: Would be nice to use a constant here.

File: consensus/types/src/chain_spec.rs
Comment (paulhauner) [inline consensus/types/src/chain_spec.rs]: Constant here too.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Looking good! I just have a few minor nits.

It might be nice to make some issues for the TODOs before this merges. I'm happy to do that, here's a list so I don't forget (feel free to add):

- Enable Web3Signer support.
- Add support for custom gas values.
- Use of timestamp value?

File: beacon_node/http_api/src/lib.rs
Comment (realbigsean) [reply beacon_node/http_api/src/lib.rs]: This exists in the `prepare_beacon_proposer`, not sure the reasoning behind it being put there, but I can remove it from both if we'd like?

File: validator_client/src/signing_method.rs
Comment (realbigsean) [reply validator_client/src/signing_method.rs:191]: I split this method into two: 

- `get_signature_from_root` -- can be used for deposits and is used for validator registrations (validator registrations also sign across the genesis fork version)
- `get_signature` -- used for everything else

So we can still sign over deposits with this method but will have to use `get_signature_from_root`. I added a check to this method to make sure no fork info is provided if those two types are being signed over.

File: consensus/types/src/application_domain.rs
Comment (realbigsean) [reply consensus/types/src/application_domain.rs]: Went with a constant here for now

**Comment by realbigsean** [comment]
> Use of timestamp value?

Made issues for the other two, but not sure I understand this one

File: beacon_node/http_api/src/lib.rs
Comment (paulhauner) [reply beacon_node/http_api/src/lib.rs]: Oh interesting, I missed this on a review from an external contributor. Let's pull them both out if you don't mind.

**Review by paulhauner — APPROVED** [review]
Awesome, feel free to bors after we pull out those IP addresses :rocket:

**Comment by realbigsean** [comment]
Already pulled them out in this commit: https://github.com/sigp/lighthouse/pull/3194/commits/a73bafeba9492cb3abaafce89bf449cc118a9923

**Pull request closed without merging**
